File: src/android/app/src/main/java/com/ijap/app/utils/CurrencyUtils.kt
package com.ijap.app.utils

import java.text.NumberFormat
import java.text.ParseException
import java.util.Currency
import java.util.Locale
import java.math.BigDecimal
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.abs

/**
 * Utility object providing comprehensive currency handling for Jewish donations
 * with support for international currencies, Israeli Shekel, and culturally significant amount handling.
 *
 * @version 1.0
 */
object CurrencyUtils {

    // Minimum donation amounts per currency
    private val MIN_DONATION_AMOUNTS = mapOf(
        "USD" to 18.0,
        "EUR" to 16.0,
        "ILS" to 60.0
    )

    // Maximum donation amounts per currency
    private val MAX_DONATION_AMOUNTS = mapOf(
        "USD" to 1000000.0,
        "EUR" to 850000.0,
        "ILS" to 3500000.0
    )

    private const val DEFAULT_CURRENCY = "USD"
    private val SUPPORTED_CURRENCIES = setOf("USD", "EUR", "ILS")
    private const val DECIMAL_PLACES = 2

    // Thread-safe cache for NumberFormat instances
    private val CURRENCY_FORMAT_CACHE = ConcurrentHashMap<String, NumberFormat>()

    /**
     * Formats a numeric amount to a localized currency string with RTL support and cultural considerations.
     *
     * @param amount The numeric amount to format
     * @param currencyCode The ISO 4217 currency code (USD, EUR, ILS)
     * @param locale Optional locale override (defaults to system locale)
     * @param useChaiNotation Whether to use chai notation for multiples of 18
     * @return Formatted currency string with proper symbol placement and RTL support
     * @throws IllegalArgumentException if currency code is not supported
     */
    @Throws(IllegalArgumentException::class)
    fun formatCurrency(
        amount: Double,
        currencyCode: String,
        locale: String? = null,
        useChaiNotation: Boolean = false
    ): String {
        require(SUPPORTED_CURRENCIES.contains(currencyCode)) {
            "Unsupported currency code: $currencyCode"
        }

        val formattingLocale = locale?.let { Locale(it) } ?: when (currencyCode) {
            "ILS" -> Locale("he", "IL")
            else -> Locale.getDefault()
        }

        val formatter = CURRENCY_FORMAT_CACHE.getOrPut(currencyCode) {
            NumberFormat.getCurrencyInstance(formattingLocale).apply {
                currency = Currency.getInstance(currencyCode)
                minimumFractionDigits = DECIMAL_PLACES
                maximumFractionDigits = DECIMAL_PLACES
            }
        }

        // Handle chai notation if requested and amount is a multiple of 18
        if (useChaiNotation && abs(amount % 18.0) < 0.001) {
            val chaiMultiple = (amount / 18.0).toInt()
            return "${formatter.format(amount)} (${chaiMultiple}×ח״י)"
        }

        return formatter.format(amount)
    }

    /**
     * Parses a currency string with support for cultural notations and multiple formats.
     *
     * @param currencyString The string to parse
     * @param currencyCode The expected currency code
     * @return Parsed amount with proper precision
     * @throws ParseException if the string cannot be parsed
     */
    @Throws(ParseException::class)
    fun parseCurrencyString(currencyString: String, currencyCode: String): Double {
        require(SUPPORTED_CURRENCIES.contains(currencyCode)) {
            "Unsupported currency code: $currencyCode"
        }

        // Remove cultural and religious notations
        val cleanString = currencyString
            .replace(Regex("\\(.*?\\)"), "") // Remove parenthetical notes
            .replace("ח״י", "") // Remove chai symbol
            .replace(Regex("[^0-9.,\\-]"), "") // Keep only numbers and decimal separator

        val formatter = CURRENCY_FORMAT_CACHE.getOrPut(currencyCode) {
            NumberFormat.getCurrencyInstance().apply {
                currency = Currency.getInstance(currencyCode)
            }
        }

        val parsedAmount = try {
            BigDecimal(cleanString).setScale(DECIMAL_PLACES, BigDecimal.ROUND_HALF_UP).toDouble()
        } catch (e: NumberFormatException) {
            throw ParseException("Unable to parse currency string: $currencyString", 0)
        }

        if (!validateAmount(parsedAmount, currencyCode)) {
            throw ParseException("Amount outside valid range for $currencyCode", 0)
        }

        return parsedAmount
    }

    /**
     * Validates donation amount against currency-specific rules and cultural requirements.
     *
     * @param amount The amount to validate
     * @param currencyCode The currency code for validation rules
     * @return True if amount meets all validation criteria
     */
    fun validateAmount(amount: Double, currencyCode: String): Boolean {
        if (!SUPPORTED_CURRENCIES.contains(currencyCode)) {
            return false
        }

        val minAmount = MIN_DONATION_AMOUNTS[currencyCode] ?: return false
        val maxAmount = MAX_DONATION_AMOUNTS[currencyCode] ?: return false

        return amount >= minAmount && 
               amount <= maxAmount && 
               BigDecimal(amount).scale() <= DECIMAL_PLACES
    }

    /**
     * Converts amounts between currencies with caching and fallback mechanisms.
     *
     * @param amount The amount to convert
     * @param fromCurrency Source currency code
     * @param toCurrency Target currency code
     * @return Converted amount with proper precision
     * @throws CurrencyConversionException if conversion fails
     */
    @Throws(CurrencyConversionException::class)
    fun convertCurrency(amount: Double, fromCurrency: String, toCurrency: String): Double {
        require(SUPPORTED_CURRENCIES.contains(fromCurrency)) {
            "Unsupported source currency: $fromCurrency"
        }
        require(SUPPORTED_CURRENCIES.contains(toCurrency)) {
            "Unsupported target currency: $toCurrency"
        }

        if (fromCurrency == toCurrency) {
            return amount
        }

        try {
            // Note: Actual exchange rate fetching would be implemented here
            // This is a simplified implementation
            val rate = when {
                fromCurrency == "USD" && toCurrency == "ILS" -> 3.5
                fromCurrency == "ILS" && toCurrency == "USD" -> 1/3.5
                fromCurrency == "EUR" && toCurrency == "USD" -> 1.1
                fromCurrency == "USD" && toCurrency == "EUR" -> 1/1.1
                fromCurrency == "EUR" && toCurrency == "ILS" -> 3.85
                fromCurrency == "ILS" && toCurrency == "EUR" -> 1/3.85
                else -> throw CurrencyConversionException("Unsupported conversion pair")
            }

            return BigDecimal(amount)
                .multiply(BigDecimal(rate))
                .setScale(DECIMAL_PLACES, BigDecimal.ROUND_HALF_UP)
                .toDouble()
        } catch (e: Exception) {
            throw CurrencyConversionException("Failed to convert currency", e)
        }
    }

    class CurrencyConversionException : Exception {
        constructor(message: String) : super(message)
        constructor(message: String, cause: Throwable) : super(message, cause)
    }
}

File: src/web/src/utils/currency.utils.ts
/**
 * Currency utility functions for the International Jewish Association Donation Platform
 * Provides comprehensive currency handling with special support for Israeli market
 * @version 1.0.0
 */

import { IDonation } from '../interfaces/donation.interface';

// Supported currencies with validation
export const SUPPORTED_CURRENCIES = ['USD', 'EUR', 'ILS'] as const;
export type SupportedCurrency = typeof SUPPORTED_CURRENCIES[number];

// Default currency for the platform
export const DEFAULT_CURRENCY: SupportedCurrency = 'USD';

// Minimum donation amounts per currency (ILS minimum set to 18 for cultural significance)
export const MIN_DONATION_AMOUNTS: Record<SupportedCurrency, number> = {
    USD: 5,
    EUR: 5,
    ILS: 18
};

// Currency symbols with security considerations
export const CURRENCY_SYMBOLS: Record<SupportedCurrency, string> = {
    USD: '$',
    EUR: '€',
    ILS: '₪'
};

// Currency-specific decimal place requirements
export const CURRENCY_DECIMAL_PLACES: Record<SupportedCurrency, number> = {
    USD: 2,
    EUR: 2,
    ILS: 0  // Israeli Shekel typically doesn't use decimal places
};

// Cache for NumberFormat instances
const formattersCache = new Map<string, Intl.NumberFormat>();

/**
 * Formats a number as a currency string with appropriate symbol and decimal places
 * Includes RTL support for Israeli Shekel
 * @param amount - The amount to format
 * @param currencyCode - The currency code (USD, EUR, ILS)
 * @returns Formatted currency string
 * @throws Error if currency is not supported
 */
export function formatCurrency(amount: number, currencyCode: string): string {
    if (!isSupportedCurrency(currencyCode)) {
        throw new Error(`Unsupported currency: ${currencyCode}`);
    }

    const cacheKey = `${currencyCode}`;
    let formatter = formattersCache.get(cacheKey);

    if (!formatter) {
        formatter = new Intl.NumberFormat('en-US', {
            style: 'currency',
            currency: currencyCode,
            minimumFractionDigits: CURRENCY_DECIMAL_PLACES[currencyCode as SupportedCurrency],
            maximumFractionDigits: CURRENCY_DECIMAL_PLACES[currencyCode as SupportedCurrency]
        });
        formattersCache.set(cacheKey, formatter);
    }

    // Add RTL markers for ILS
    const formattedAmount = formatter.format(amount);
    return currencyCode === 'ILS' ? `\u200e${formattedAmount}\u200f` : formattedAmount;
}

/**
 * Validates if an amount meets the currency-specific requirements
 * @param amount - The amount to validate
 * @param currencyCode - The currency code
 * @returns boolean indicating if amount is valid
 * @throws Error if currency is not supported
 */
export function validateCurrencyAmount(amount: number, currencyCode: string): boolean {
    if (!isSupportedCurrency(currencyCode)) {
        throw new Error(`Unsupported currency: ${currencyCode}`);
    }

    const typedCurrency = currencyCode as SupportedCurrency;

    // Basic amount validation
    if (!Number.isFinite(amount) || amount <= 0) {
        return false;
    }

    // Check minimum amount
    if (amount < MIN_DONATION_AMOUNTS[typedCurrency]) {
        return false;
    }

    // Validate decimal places
    const decimalPlaces = (amount.toString().split('.')[1] || '').length;
    if (decimalPlaces > CURRENCY_DECIMAL_PLACES[typedCurrency]) {
        return false;
    }

    // Check for numeric overflow
    if (amount > Number.MAX_SAFE_INTEGER) {
        return false;
    }

    return true;
}

/**
 * Returns the symbol for a given currency code
 * @param currencyCode - The currency code
 * @returns The currency symbol
 * @throws Error if currency is not supported
 */
export function getCurrencySymbol(currencyCode: string): string {
    if (!isSupportedCurrency(currencyCode)) {
        throw new Error(`Unsupported currency: ${currencyCode}`);
    }

    // Sanitize and return the currency symbol
    return CURRENCY_SYMBOLS[currencyCode as SupportedCurrency];
}

/**
 * Returns the minimum donation amount for a given currency
 * @param currencyCode - The currency code
 * @returns The minimum donation amount
 * @throws Error if currency is not supported
 */
export function getMinimumAmount(currencyCode: string): number {
    if (!isSupportedCurrency(currencyCode)) {
        throw new Error(`Unsupported currency: ${currencyCode}`);
    }

    return MIN_DONATION_AMOUNTS[currencyCode as SupportedCurrency];
}

/**
 * Checks if a currency code is supported by the platform
 * @param currencyCode - The currency code to check
 * @returns boolean indicating if currency is supported
 */
export function isSupportedCurrency(currencyCode: string): boolean {
    if (!currencyCode || typeof currencyCode !== 'string') {
        return false;
    }

    const sanitizedCode = currencyCode.toUpperCase().trim();
    return SUPPORTED_CURRENCIES.includes(sanitizedCode as SupportedCurrency);
}

/**
 * Type guard for SupportedCurrency
 * @param currency - The currency to check
 * @returns Type predicate for SupportedCurrency
 */
export function isSupportedCurrencyType(currency: string): currency is SupportedCurrency {
    return isSupportedCurrency(currency);
}